android/src/main/java/com/reactnativepagerview/FragmentAdapter.kt

```diff
@@ -3,54 +3,98 @@ package com.reactnativepagerview
-  private val childrenViews: MutableList<View> = ArrayList()
+  private val childrenViews = mutableListOf<View>()
+  private var count = 0
+  private var offset = 0
+  private var isDirty = false
+  private val prevItemIds = mutableListOf<Long>()
+
   override fun createFragment(position: Int): Fragment {
-    return ViewPagerFragment(childrenViews[position])
+    return ViewPagerFragment(getViewAtPosition(position))
   }
 
-  override fun getItemCount(): Int {
-    return childrenViews.size
-  }
+  override fun getItemCount() = count
 
   override fun getItemId(position: Int): Long {
-    return childrenViews[position].id.toLong()
+    val view = getViewAtPosition(position)
+    return view?.id?.toLong() ?: UNRENDERED_ID_OFFSET + position
   }
 
   override fun containsItem(itemId: Long): Boolean {
-    for (child in childrenViews) {
-      if (itemId.toInt() == child.id) {
-        return true
-      }
+    if (itemId >= UNRENDERED_ID_OFFSET) {
+      val position = itemId - UNRENDERED_ID_OFFSET
+      return getViewAtPosition(position.toInt()) == null
     }
-    return false
+    return childrenViews.any { it.id.toLong() == itemId }
   }
 
-  fun addFragment(child: View, index: Int) {
-    childrenViews.add(index, child)
-    notifyItemInserted(index)
+  /**
+   * Returns true if any changes were applied.
+   */
+  fun applyChanges(): Boolean {
+    if (!isDirty) {
+      return false
+    }
+
+    isDirty = false
+    val diff = DiffUtil.calculateDiff(
+      PagerDiffCallback(prevItemIds, this),
+      false
+    )
+    diff.dispatchUpdatesTo(this)
+    return true
+  }
+
+  fun setCount(count: Int) {
+    if (this.count != count) {
+      markDirty()
+      this.count = count
+    }
+  }
+
+  fun setOffset(offset: Int) {
+    if (this.offset != offset) {
+      markDirty()
+      this.offset = offset
+    }
   }
 
-  fun removeFragment(child: View) {
-    val index = childrenViews.indexOf(child)
-    removeFragmentAt(index)
+  fun addReactView(child: View, index: Int) {
+    markDirty()
+    childrenViews.add(index, child)
   }
 
-  fun removeFragmentAt(index: Int) {
+  fun getReactChildAt(index: Int) = childrenViews[index]
+
+  fun getReactChildCount() = childrenViews.size
+
+  fun removeReactViewAt(index: Int) {
+    markDirty()
     childrenViews.removeAt(index)
-    notifyItemRemoved(index)
   }
 
-  fun removeAll() {
-    childrenViews.clear()
-    notifyDataSetChanged()
+  private fun getViewAtPosition(position: Int): View? {
+    val index = position - offset
+    return if (index >= 0 && index < childrenViews.size) childrenViews[index] else null
+  }
+
+  private fun markDirty() {
+    if (isDirty) {
+      return
+    }
+    isDirty = true
+    prevItemIds.clear()
+    for (position in 0 until itemCount) {
+      prevItemIds.add(getItemId(position))
+    }
   }
 
-  fun getChildViewAt(index: Int): View {
-    return childrenViews[index]
+  companion object {
+     const val UNRENDERED_ID_OFFSET = 0xffffffffL
```

Comment:
Could you explain me the purpose of this static val ?

Reply:
Every page in `ViewPager2` must have an id. The id must stably refer to the same page, otherwise data mutations could cause the pager to lose track of what is the current page.

The simplest way is to use the page's `View` id (this is how the current master branch works). RN manages the id stability.

However with lazy rendering, a specific page may not have a `View` available when asked to provide an id.  So instead (when the `View` is not available), compute the id as `UNRENDERED_ID_OFFSET + pageIndex`.  This guarantees a unique id since all `View` ids are less than `UNRENDERED_ID_OFFSET`.  The custom `PagerDiffCallback` uses this constant as a threshold to determine if a page is rendered, so when a new render arrives, the pager can swap the unrendered for rendered without losing track of what the current page is.

Reply:
Thank you for an explanation :D